OpenRA.Game/RayCaster.cs

```diff
@@ -0,0 +1,188 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using OpenRA.Traits;
+
+namespace OpenRA
+{
+	// Algorithm obtained from ftp://ftp.isc.org/pub/usenet/comp.sources.unix/volume26/line3d
+	public class Raycaster
+	{
+		public static IEnumerable<WPos> Raycast(WPos source, WPos target)
+		{
+			var xDelta = target.X - source.X;
+			var yDelta = target.Y - source.Y;
+			var zDelta = target.Z - source.Z;
+
+			var xAbsolute = Math.Abs(xDelta) << 1;
+			var yAbsolute = Math.Abs(yDelta) << 1;
+			var zAbsolute = Math.Abs(zDelta) << 1;
+
+			var xIncrement = (xDelta < 0) ? -1 : xDelta > 0 ? 1 : 0;
+			var yIncrement = (yDelta < 0) ? -1 : yDelta > 0 ? 1 : 0;
+			var zIncrement = (zDelta < 0) ? -1 : zDelta > 0 ? 1 : 0;
+
+			var x = source.X;
+			var y = source.Y;
+			var z = source.Z;
+
+			if (xAbsolute >= Math.Max(yAbsolute, zAbsolute))
+			{
+				var yError = yAbsolute - (xAbsolute >> 1);
+				var zError = zAbsolute - (xAbsolute >> 1);
+				for (;;)
```

Comment:
I understand forever loops are frowned upon, so if someone could help me extract the function here I'd be grateful.